File: foreign/go/internal/vsr/namespace.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package vsr

import (
	"encoding/binary"

	ierror "github.com/apache/iggy/foreign/go/errors"
	"github.com/apache/iggy/foreign/go/internal/command"
)

// Namespace packing limits, a port of
// core/binary_protocol/src/namespace.rs.
const (
	MaxStreams    = 4096
	MaxTopics     = 4096
	MaxPartitions = 1_000_000

	streamBits    = 12
	topicBits     = 12
	partitionBits = 20

	partitionShift = 0
	topicShift     = partitionShift + partitionBits
	streamShift    = topicShift + topicBits
)

// MetadataConsensusNamespace routes a control-plane request to the metadata
// replica on shard 0. Plain 0 falls into namespace hashing and can land a
// Register on a peer shard that has no consensus instance.
const MetadataConsensusNamespace uint64 = 1 << 63

// Identifier kinds in the wire prefix [kind u8][len u8][value].
const (
	identifierKindNumeric = 1
	identifierKindString  = 2
)

// partitioningPartitionID is the only Partitioning kind that carries an
// explicit partition, which is what routing needs.
const partitioningPartitionID = 2

// PackNamespace packs stream, topic and partition ids into a routing
// namespace.
func PackNamespace(streamID, topicID, partitionID uint32) (uint64, error) {
	if err := validateNamespaceField(streamID, MaxStreams); err != nil {
		return 0, err
	}
	if err := validateNamespaceField(topicID, MaxTopics); err != nil {
		return 0, err
	}
	if err := validateNamespaceField(partitionID, MaxPartitions); err != nil {
		return 0, err
	}
	return uint64(streamID)<<streamShift |
		uint64(topicID)<<topicShift |
		uint64(partitionID)<<partitionShift, nil
}

// NamespaceForRequest selects the routing namespace for a request.
// Partition-plane commands derive it by peeking their own encoded payload. A
// named stream or topic identifier yields 0 so the server resolves the name.
func NamespaceForRequest(code uint32, payload []byte, operation Operation) (uint64, error) {
	if operation == OperationRegister || operation == OperationLogout {
		return MetadataConsensusNamespace, nil
	}
	if operation == OperationNonReplicated || IsMetadata(operation) {
		return 0, nil
	}

	switch command.Code(code) {
	case command.SendMessagesCode:
		return namespaceFromSendMessages(payload)
	case command.StoreOffsetCode, command.DeleteConsumerOffsetCode,
		command.StoreOffset2Code, command.DeleteConsumerOffset2Code:
		return namespaceFromConsumerOffset(payload)
	case command.DeleteSegmentsCode:
		return namespaceFromDeleteSegments(payload)
	default:
		// The guard that keeps a partition-classified operation this SDK
		// cannot peek from reaching the wire with a wrong namespace.
		return 0, ierror.ErrFeatureUnavailable
	}
}

// peekedIdentifier is a decoded identifier prefix. numeric is zero and named
// is true for a string identifier, which the server resolves.
type peekedIdentifier struct {
	numeric uint32
	named   bool
	length  int
}

func peekIdentifier(payload []byte, offset int) (peekedIdentifier, error) {
	if len(payload) < offset+2 {
		return peekedIdentifier{}, ierror.ErrInvalidCommand
	}
	kind := payload[offset]
	length := int(payload[offset+1])
	if len(payload) < offset+2+length {
		return peekedIdentifier{}, ierror.ErrInvalidCommand
	}
	switch {
	case kind == identifierKindNumeric && length == 4:
		return peekedIdentifier{
			numeric: binary.LittleEndian.Uint32(payload[offset+2:]),
			length:  2 + length,
		}, nil
	case kind == identifierKindString && length > 0:
		return peekedIdentifier{named: true, length: 2 + length}, nil
	default:
		return peekedIdentifier{}, ierror.ErrInvalidCommand
	}
}

func validateNamespaceField(value uint32, exclusiveMax uint32) error {
	if value >= exclusiveMax {
		return ierror.ErrInvalidIdentifier
	}
	return nil
}

func namespaceFromIdentifiers(stream, topic peekedIdentifier, partitionID uint32) (uint64, error) {
	if stream.named || topic.named {
		return 0, nil
	}
	return PackNamespace(stream.numeric, topic.numeric, partitionID)
}

// namespaceFromSendMessages peeks
// [metadata_len u32][stream ident][topic ident][partitioning kind u8, len u8, value].
// Only explicit PartitionId partitioning is routable: the broker never picks a
// partition under consensus.
func namespaceFromSendMessages(payload []byte) (uint64, error) {
	if len(payload) < 4 {
		return 0, ierror.ErrInvalidCommand
	}
	metadataLength := uint64(binary.LittleEndian.Uint32(payload))
	if uint64(len(payload)) < 4+metadataLength {
		return 0, ierror.ErrInvalidCommand
	}
	// A read past the declared metadata region must fail rather than spill
	// into message bytes and derive a namespace the server would not compute.
	metadata := payload[4 : 4+metadataLength]

	offset := 0
	stream, err := peekIdentifier(metadata, offset)
	if err != nil {
		return 0, err
	}
	offset += stream.length
	topic, err := peekIdentifier(metadata, offset)
	if err != nil {
		return 0, err
	}
	offset += topic.length

	if len(metadata) < offset+2 {
		return 0, ierror.ErrInvalidCommand
	}
	partitioningKind := metadata[offset]
	partitioningLength := int(metadata[offset+1])
	if partitioningKind != partitioningPartitionID {
		return 0, ierror.ErrFeatureUnavailable
	}
	if partitioningLength != 4 || len(metadata) < offset+2+4 {
		return 0, ierror.ErrInvalidCommand
	}
	partitionID := binary.LittleEndian.Uint32(metadata[offset+2:])

	return namespaceFromIdentifiers(stream, topic, partitionID)
}

// namespaceFromConsumerOffset peeks
// [consumer kind u8][consumer ident][stream ident][topic ident]
// [partition flag u8][partition u32]. The v1 and v2 request layouts share this
// prefix and differ only in the trailing fields, which routing ignores.
func namespaceFromConsumerOffset(payload []byte) (uint64, error) {
	if len(payload) < 1 || (payload[0] != 1 && payload[0] != 2) {
		return 0, ierror.ErrInvalidCommand
	}
	offset := 1
	consumer, err := peekIdentifier(payload, offset)
	if err != nil {
		return 0, err
	}
	offset += consumer.length
	stream, err := peekIdentifier(payload, offset)
	if err != nil {
		return 0, err
	}
	offset += stream.length
	topic, err := peekIdentifier(payload, offset)
	if err != nil {
		return 0, err
	}
	offset += topic.length

	if len(payload) < offset+5 {
		return 0, ierror.ErrInvalidCommand
	}
	if payload[offset] != 1 {
		return 0, ierror.ErrInvalidIdentifier
	}
	partitionID := binary.LittleEndian.Uint32(payload[offset+1:])

	return namespaceFromIdentifiers(stream, topic, partitionID)
}

// namespaceFromDeleteSegments peeks
// [stream ident][topic ident][partition u32].
func namespaceFromDeleteSegments(payload []byte) (uint64, error) {
	offset := 0
	stream, err := peekIdentifier(payload, offset)
	if err != nil {
		return 0, err
	}
	offset += stream.length
	topic, err := peekIdentifier(payload, offset)
	if err != nil {
		return 0, err
	}
	offset += topic.length

	if len(payload) < offset+4 {
		return 0, ierror.ErrInvalidCommand
	}
	partitionID := binary.LittleEndian.Uint32(payload[offset:])

	return namespaceFromIdentifiers(stream, topic, partitionID)
}

File: foreign/node/src/wire/vsr/namespace.ts
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.
//

/**
 * Namespace packing and the partition-plane payload peeks needed to derive
 * it, ported from `core/binary_protocol/src/namespace.rs` and
 * `namespace_for_request` in `core/sdk/src/vsr.rs`.
 */

import { COMMAND_CODE } from '../command.code.js';
import { responseError } from '../error.utils.js';
import { Operation, isMetadata } from './operation.js';

/** `IggyError::InvalidCommand`. */
const INVALID_COMMAND = 3;
/** `IggyError::FeatureUnavailable`. */
const FEATURE_UNAVAILABLE = 5;
/** `IggyError::InvalidIdentifier`. */
const INVALID_IDENTIFIER = 6;

const MAX_STREAMS = 4096;
const MAX_TOPICS = 4096;
const MAX_PARTITIONS = 1_000_000;
const bitsRequired = (value: number): bigint =>
  BigInt(BigInt(value).toString(2).length);
const TOPIC_SHIFT = bitsRequired(MAX_PARTITIONS - 1);
const STREAM_SHIFT = TOPIC_SHIFT + bitsRequired(MAX_TOPICS - 1);

/**
 * Control-plane requests target the metadata replica (shard 0), selected by
 * this exact sentinel. Plain 0 would fall into namespace hashing and land a
 * Register on a peer shard.
 */
export const METADATA_CONSENSUS_NAMESPACE = 1n << 63n;

/** Packs stream / topic / partition ids into a routing namespace. */
export const packNamespace = (
  streamId: number,
  topicId: number,
  partitionId: number
): bigint => {
  validateField(streamId, MAX_STREAMS);
  validateField(topicId, MAX_TOPICS);
  validateField(partitionId, MAX_PARTITIONS);
  return (BigInt(streamId) << STREAM_SHIFT) |
    (BigInt(topicId) << TOPIC_SHIFT) |
    BigInt(partitionId);
};

/**
 * Selects the routing namespace for a request. Partition-plane commands
 * derive it from their own payload; a named stream or topic identifier
 * yields 0 so the server resolves the name.
 *
 * @throws Error mirroring the Rust SDK: invalid-identifier for an
 *   out-of-range field, invalid-command for an undecodable payload, and
 *   feature-unavailable for a partition operation this SDK cannot derive.
 */
export const namespaceForRequest = (
  code: number,
  payload: Buffer,
  operation: number
): bigint => {
  if (operation === Operation.Register || operation === Operation.Logout)
    return METADATA_CONSENSUS_NAMESPACE;
  if (operation === Operation.NonReplicated || isMetadata(operation))
    return 0n;

  switch (code) {
    case COMMAND_CODE.SendMessages:
      return namespaceFromSendMessages(payload);
    case COMMAND_CODE.StoreOffset:
    case COMMAND_CODE.DeleteConsumerOffset:
    case COMMAND_CODE.StoreOffset2:
    case COMMAND_CODE.DeleteConsumerOffset2:
      return namespaceFromConsumerOffset(payload);
    case COMMAND_CODE.DeleteSegments:
      return namespaceFromDeleteSegments(payload);
    default:
      // The guard that keeps custom partition operations unreachable.
      throw responseError(code, FEATURE_UNAVAILABLE);
  }
};

/** A decoded identifier: numeric value, or null for a name (server resolves). */
type PeekedIdentifier = {
  numeric: number | null,
  length: number
};

const IDENTIFIER_KIND_NUMERIC = 1;
const IDENTIFIER_KIND_STRING = 2;

const peekIdentifier = (payload: Buffer, offset: number): PeekedIdentifier => {
  if (payload.length < offset + 2)
    throw responseError(0, INVALID_COMMAND);
  const kind = payload.readUInt8(offset);
  const length = payload.readUInt8(offset + 1);
  if (payload.length < offset + 2 + length)
    throw responseError(0, INVALID_COMMAND);
  if (kind === IDENTIFIER_KIND_NUMERIC) {
    if (length !== 4)
      throw responseError(0, INVALID_COMMAND);
    return { numeric: payload.readUInt32LE(offset + 2), length: 2 + length };
  }
  if (kind === IDENTIFIER_KIND_STRING && length > 0)
    return { numeric: null, length: 2 + length };
  throw responseError(0, INVALID_COMMAND);
};

const validateField = (value: number, exclusiveMax: number): void => {
  if (!Number.isInteger(value) || value < 0 || value >= exclusiveMax)
    throw responseError(0, INVALID_IDENTIFIER);
};

const namespaceFromIds = (
  stream: PeekedIdentifier,
  topic: PeekedIdentifier,
  partitionId: number
): bigint => {
  // Named identifiers defer resolution to the server.
  if (stream.numeric === null || topic.numeric === null) return 0n;
  return packNamespace(stream.numeric, topic.numeric, partitionId);
};

/**
 * `SendMessages`: `[metadata_len u32][stream ident][topic ident]
 * [partitioning kind u8, len u8, value]...`. Only explicit `PartitionId`
 * partitioning is routable under VSR; the broker never picks a partition.
 * TODO(hubcio): Balanced and MessageKey partitioning to be implemented;
 * not decided yet whether it'll be on server side or client side.
 */
const namespaceFromSendMessages = (payload: Buffer): bigint => {
  if (payload.length < 4)
    throw responseError(COMMAND_CODE.SendMessages, INVALID_COMMAND);
  const metadataLength = payload.readUInt32LE(0);
  if (payload.length < 4 + metadataLength)
    throw responseError(COMMAND_CODE.SendMessages, INVALID_COMMAND);
  // Rust peeks inside payload[4..4 + metadata_length]; a read past the
  // declared metadata region must fail rather than spill into message bytes
  // and derive a namespace the server would never compute.
  const metadata = payload.subarray(4, 4 + metadataLength);

  let offset = 0;
  const stream = peekIdentifier(metadata, offset);
  offset += stream.length;
  const topic = peekIdentifier(metadata, offset);
  offset += topic.length;

  if (metadata.length < offset + 2)
    throw responseError(COMMAND_CODE.SendMessages, INVALID_COMMAND);
  const partitioningKind = metadata.readUInt8(offset);
  const partitioningLength = metadata.readUInt8(offset + 1);
  const PARTITIONING_PARTITION_ID = 2;
  if (partitioningKind !== PARTITIONING_PARTITION_ID)
    throw responseError(COMMAND_CODE.SendMessages, FEATURE_UNAVAILABLE);
  if (partitioningLength !== 4 || metadata.length < offset + 2 + 4)
    throw responseError(COMMAND_CODE.SendMessages, INVALID_COMMAND);
  const partitionId = metadata.readUInt32LE(offset + 2);

  return namespaceFromIds(stream, topic, partitionId);
};

/**
 * Consumer-offset requests: `[consumer kind u8][consumer ident]
 * [stream ident][topic ident][partition flag u8][partition u32]...`.
 */
const namespaceFromConsumerOffset = (payload: Buffer): bigint => {
  if (payload.length < 1 || (payload.readUInt8(0) !== 1 &&
      payload.readUInt8(0) !== 2))
    throw responseError(COMMAND_CODE.StoreOffset, INVALID_COMMAND);
  let offset = 1;
  const consumer = peekIdentifier(payload, offset);
  offset += consumer.length;
  const stream = peekIdentifier(payload, offset);
  offset += stream.length;
  const topic = peekIdentifier(payload, offset);
  offset += topic.length;

  if (payload.length < offset + 5)
    throw responseError(COMMAND_CODE.StoreOffset, INVALID_COMMAND);
  const hasPartition = payload.readUInt8(offset) === 1;
  if (!hasPartition)
    throw responseError(COMMAND_CODE.StoreOffset, INVALID_IDENTIFIER);
  const partitionId = payload.readUInt32LE(offset + 1);

  return namespaceFromIds(stream, topic, partitionId);
};

/** `DeleteSegments`: `[stream ident][topic ident][partition u32]...`. */
const namespaceFromDeleteSegments = (payload: Buffer): bigint => {
  let offset = 0;
  const stream = peekIdentifier(payload, offset);
  offset += stream.length;
  const topic = peekIdentifier(payload, offset);
  offset += topic.length;

  if (payload.length < offset + 4)
    throw responseError(COMMAND_CODE.DeleteSegments, INVALID_COMMAND);
  const partitionId = payload.readUInt32LE(offset);

  return namespaceFromIds(stream, topic, partitionId);
};
